### PR TITLE
Tracking doc: rewrite `@wdio/allure-reporter` package into Typescript

### LINKED ISSUE #5825: Tracking doc: rewrite `@wdio/allure-reporter` package into TypeScript
This is a tracking issue for organising the effort of re-writing the code base from vanilla JS into TypeScript. If you are reading this and are interested getting involved, this is a great chance to do so. Here are the rules:

- if you are interested converting source files from `/packages/wdio-allure-reporter` into TypeScript please comment on this thread and mention the files you want to convert to avoid duplicate work
- there is no reason to convert the whole package at once - smaller PRs make it easier to review the code
- if you make a PR, reference this issue so we can keep track of the progress
- if you make a PR, label it with `TypeScript` and the package name `wdio-allure-reporter` so we can easier find all PRs
- if you make a PR, assign it to the `TypeScript Rewrite` project

Happy rewriting 😅

### PR DESCRIPTION
**Pull request by hieuxlu**
Closes #5825

## Proposed changes

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [X] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have added necessary documentation (if appropriate)
- [X] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

File: packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts:34]:
```suggestion
        constructor();
        setOptions(options: Allure.Options): void;
        getCurrentSuite(): Allure.Suite;
        getCurrentTest(): Allure.Test;
        startSuite(suiteName: string, timestamp?: number): void;
        endSuite(timestamp?: number): void;
        startCase(testName: string, timestamp?: number): void;
        endCase(status: Allure.Status, err?: {}, timestamp?: number): void;
        startStep(stepName: string, timestamp?: number): void;
        endStep(status: Allure.Status, timestamp?: number): void;
        setDescription(description: string, timestamp?: number): void;
        addAttachment(attachmentName: string, buffer: any, type?: string): void;
        pendingCase(testName: string, timestamp?: number): void;
```

File: packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts:118]: can we clean up the empty lines between function definition within the classes? It doesn't follow other TS declarations and bloats up the file a bit

File: packages/wdio-allure-reporter/src/constants.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/src/constants.ts:27]:
```suggestion
const events: Record<string, string> = {

```

File: packages/wdio-allure-reporter/src/utils.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/src/utils.ts:63]:
```suggestion
export const tellReporter = (event: string, msg: any = {}) => {

```

File: packages/wdio-allure-reporter/tests/__fixtures__/command.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/tests/__fixtures__/command.ts:36]:
```suggestion
export function commandStart(isDevtools: boolean) {

```

File: packages/wdio-allure-reporter/tests/__fixtures__/command.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/tests/__fixtures__/command.ts:40]:
```suggestion
export function commandEnd(isDevtools: boolean) {

```

File: packages/wdio-allure-reporter/tests/__fixtures__/command.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/tests/__fixtures__/command.ts:44]:
```suggestion
export function commandStartScreenShot(isDevtools: boolean) {

```

File: packages/wdio-allure-reporter/tests/__fixtures__/command.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/tests/__fixtures__/command.ts:48]:
```suggestion
export function commandEndScreenShot(isDevtools: boolean) {

```

**Review by christian-bromann — CHANGES_REQUESTED** [review]
It seems you removed the Allure type generator which was used to autogenerate the Allure reporter type definition. I am all for it to get rid of this but we need to ensure that it is documented that changes to the class (adding a static method) needs to be documented in the type definition as well as in the docs.

File: packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts
Comment (hieuxlu) [reply packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts:118]: This is the original file from https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/allure-js-commons
I added the reason not to use the types package on top. I wonder if it's worth it to refactor all those spaces.
```
// Allure is providing its own types in v2, but there's too many changes,
// so this is temporarily copied here to customize typings with allure 1
```

File: packages/wdio-allure-reporter/src/constants.ts
Comment (hieuxlu) [reply packages/wdio-allure-reporter/src/constants.ts:27]: This does not work because Node 12 hard coded the event string for `process.on()`. Casting to any is the only way now.

File: packages/wdio-allure-reporter/src/utils.ts
Comment (hieuxlu) [reply packages/wdio-allure-reporter/src/utils.ts:63]: This does not work because Node 12 hard coded the event string for `process.emit()`. Casting to any is the only way now.

**Comment by hieuxlu** [comment]
Typedefs for `@wdio/reporter` is really messed up. `index.ts` does not map with `reporter.d.ts`. This PR is gonna need some more times.

File: packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts
Comment (christian-bromann) [reply packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts:118]: If it is maintained in definitely types can't we just import it?

Comment (BorisOsipov) [reply packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts:118]: @christian-bromann allure1 is obsolete and isn't maintained. So I see reasonable to copy `d.ts` if it makes easier to move wdio-allure to ts.

Comment (hieuxlu) [reply packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts:118]: The types in Definitely Types are not ideal, I can't use it without customization.

**Comment by christian-bromann** [comment]
@hieuxlu adding a work-in-progress label. Let me know if you need any help or another review. Thanks!

File: packages/wdio-allure-reporter/src/index.ts
Comment (hieuxlu) [inline packages/wdio-allure-reporter/src/index.ts:8]: @christian-bromann Added this. The previous generated typings does not automatically update the docs in README.md, so this is not any worse

**Comment by hieuxlu** [comment]
@christian-bromann I ran into a type error because there are multiple `@types/node` packages resolved by lerna and npm. Did you ever encounter this problem?
```
:../../../packages/wdio-reporter/reporter.d.ts:23:2311 - 62 - error: interface Element { }
error TS4090: 9 TS4090: Conflicting definitions for 'node' found at 'D:/repos/webdriverio/packages/webdriver/node_modules/@types/node/index.d.ts' and 'D:/repos/webdriv
      Conflicting definitions for 'node' found at 'D:/repos/webdriverio/packages/webdriver/node_modules/@types/node/index.d.ts' and 'D:/repos/webdriverio/perio/packages/webdriverio/node_modules/@types/node/index.d.ts'. Consider installing a specific version of this library to resolve the conflict.
 ackages/webdriverio/node_modules/@types/node/index.d.ts'. Consider installing a specific version of this library to resolve the conflict.
```

File: .eslintrc.js
Comment (christian-bromann) [inline .eslintrc.js:43]: Why is this needed?

File: packages/wdio-allure-reporter/package.json
Comment (christian-bromann) [inline packages/wdio-allure-reporter/package.json:12]: Can you verify that the generated types are backwards compatible?

File: packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts:13]: Why do we need this if we have `@types/allure-js-commons` as dependency?

File: packages/wdio-allure-reporter/src/constants.ts
Comment (christian-bromann) [reply packages/wdio-allure-reporter/src/constants.ts:25]: I would prefer to cast it to any in the `process.on` call. This way we can ensure type safety here and just ignore it there.

File: packages/wdio-allure-reporter/src/utils.ts
Comment (christian-bromann) [reply packages/wdio-allure-reporter/src/utils.ts:64]: same, let's cast it in the `process.emit` call but keep type safety for our function.

File: tests/typings/sync/tsconfig.json
Comment (christian-bromann) [inline tests/typings/sync/tsconfig.json:5]: let's remove this, I don't see a need for it.
```suggestion
```

File: tests/typings/webdriverio/tsconfig.json
Comment (christian-bromann) [inline tests/typings/webdriverio/tsconfig.json:6]: same
```suggestion
```

**Review by christian-bromann — COMMENTED** [review]
A couple more comments

**Comment by christian-bromann** [comment]
> Did you ever encounter this problem?

@hieuxlu no. `@types/node` is in none of the sub package dependencies. It also doesn't seem to be an issue with the master branch. Maybe try to run `npm run setup-full` to re-initiate the workspace.

File: packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts
Comment (christian-bromann) [reply packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts:13]: Based on the convo above it seems that the typings are not maintained. So I am fine to copy&paste them here but then we should get rid of the `@types/allure-js-commons` dependency.

**Comment by hieuxlu** [comment]
> > Did you ever encounter this problem?
> 
> @hieuxlu no. `@types/node` is in none of the sub package dependencies. It also doesn't seem to be an issue with the master branch. Maybe try to run `npm run setup-full` to re-initiate the workspace.

I did, that's why I found it very strange

File: packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts
Comment (hieuxlu) [reply packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts:13]: Typescript will throw an error in `test:typings` if there'no `@types/allure-js-commons` packages. But the package alone does not provide enough typings, so I needed to override. I will see what is the minimum overriding needed

File: packages/wdio-allure-reporter/package.json
Comment (hieuxlu) [reply packages/wdio-allure-reporter/package.json:12]: Is the test in `test:typings` enough? Should I verify with a standalone project as well?

File: tests/typings/sync/tsconfig.json
Comment (hieuxlu) [reply tests/typings/sync/tsconfig.json:5]: This was the result of Typescript requiring `@types/allure-js-commons`.

File: packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts
Comment (christian-bromann) [reply packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts:13]: Please check why the typings test is failing. There should be no need to have `@types/allure-js-commons` if we define it ourselves.

Comment (hieuxlu) [reply packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts:13]: > Please check why the typings test is failing. There should be no need to have `@types/allure-js-commons` if we define it ourselves.

You're right, I forgot to mention somehow we're not copying .d.ts files in src folder. So I'll need to modify the build scripts a bit.

File: packages/wdio-allure-reporter/package.json
Comment (hieuxlu) [reply packages/wdio-allure-reporter/package.json:12]: @christian-bromann This causes the conflicting node types error, because these 2 lines are autogenerated by tsc
```
/// <reference types="webdriver" />
/// <reference types="node" />
import WDIOReporter from '@wdio/reporter';
import Allure from 'allure-js-commons';
```

I believe on master, we've already have conflicting `@types/node`, i.e. multiple `@types/node` packages hoisted after lerna bootstrap. It is a timebomb that will explode when we get rid of providing `.d.ts` typings manually and ship the declaration generated by Typescript. 

In order to avoid this error, I will need to revert back to manual `.d.ts` file as main file for now.

File: packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts
Comment (hieuxlu) [reply packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts:13]: I can not find any way to ship our defined `allure-js-commons.d.ts`. The problem is even if we copy it in the build folder, Typescript will not recognize the types because: 
- It is not referenced in package.json
- It does not exist in `node_modules/@types`

This is why I needed to include `@types/allure-js-commons` in the first place. It could be used as a work around for Typescript typings.

File: packages/wdio-allure-reporter/package.json
Comment (hieuxlu) [reply packages/wdio-allure-reporter/package.json:12]: Another solution is we could upgrade to Allure 2 and use its typings instead.

File: packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts
Comment (christian-bromann) [reply packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts:13]: I don't understand. There are two ways how 3rd party module types can be embedded. One is via installing `@types/allure-js-commons` and the other is to declare a module called `allure-js-commons` in a customer types file, e.g. like we do with the [lighthouse module](https://github.com/webdriverio/webdriverio/blob/master/packages/wdio-devtools-service/src/@types/lighthouse.ts#L0-L1) in the devtools-service package. So if we define a custom `allure-js-commons.d.ts` we don't need `@types/allure-js-commons`.

**Comment by christian-bromann** [comment]
@hieuxlu please let me know if I can assist in anything.

**Comment by hieuxlu** [comment]
There's been a major rewrite of `@wdio/reporter` merged last week, so I will need to rework most of the imported types

**Comment by hieuxlu** [comment]
@christian-bromann Coverage tests passed locally, but I'm blocked by #6193, because I can't get pass typing tests. I can resolve the remaining comments, but still blocked by node typing conflict.

**Comment by christian-bromann** [comment]
@hieuxlu I suggest in order to move forward with this to revert `"types": "./build/index.d.ts",` in the package.json. I will look into cleaning up old TS artefacts once we've moved all packages.

File: .eslintrc.js
Comment (christian-bromann) [reply .eslintrc.js:43]: can you comment?

File: packages/wdio-allure-reporter/package.json
Comment (christian-bromann) [reply packages/wdio-allure-reporter/package.json:12]: let's revert this change to unblock us and pick it up at a later stage.

File: packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts
Comment (christian-bromann) [reply packages/wdio-allure-reporter/src/@types/allure-js-commons.d.ts:13]: Let's either use `@types/allure-js-commons` as dependency or type this package manually.

**Comment by christian-bromann** [comment]
@hieuxlu made some more comments. I think we are at a very good state here, just need some last improvements. Happy to assist via pair programming if you like.

File: packages/wdio-reporter/src/stats/suite.ts
Comment (hieuxlu) [reply packages/wdio-reporter/src/stats/suite.ts:1]: It expect object with name and stuffs, but yeah, we can have our own-defined type here.

**Comment by hieuxlu** [comment]
> @hieuxlu made some more comments. I think we are at a very good state here, just need some last improvements. Happy to assist via pair programming if you like.

Don't worry, if I revert package.json typings property, it's no longer blocked. Will try to fix remaining comments.

File: packages/wdio-allure-reporter/src/index.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/src/index.ts:54]: We started to make all props private.

```suggestion
    private _allure: Allure;
    private _capabilities: WebDriver.DesiredCapabilities;
    private _config: WebdriverIO.Config;
    private _isMultiremote?: boolean;
    private _lastScreenshot?: string;
    private _options: AllureReporterOptions;
```

File: packages/wdio-allure-reporter/src/index.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/src/index.ts:70]:
```suggestion
```

File: packages/wdio-allure-reporter/src/index.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/src/index.ts:64]: we can use default values when defining them in the header or get rid of this entirely.

```suggestion
```

File: packages/wdio-allure-reporter/src/index.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/src/index.ts:210]:
```suggestion
    onTestFail(test: TestStats) {

```

File: packages/wdio-allure-reporter/src/index.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/src/index.ts:146]:
```suggestion
    onTestStart(test: TestStats) {

```

File: packages/wdio-allure-reporter/src/index.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/src/index.ts:89]:
```suggestion
    onRunnerStart(runner: RunnerStats) {

```

File: packages/wdio-allure-reporter/allure-reporter.d.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/allure-reporter.d.ts:1]:
```suggestion
```

File: packages/wdio-allure-reporter/src/index.ts
Comment (hieuxlu) [reply packages/wdio-allure-reporter/src/index.ts:211]: Our code currently use this method for both tests and hooks, so we need to override type here

File: packages/wdio-allure-reporter/src/index.ts
Comment (hieuxlu) [reply packages/wdio-allure-reporter/src/index.ts:147]: Our code currently use this method for both tests and hooks, so we need to override type here

Comment (christian-bromann) [reply packages/wdio-allure-reporter/src/index.ts:147]: Can you point out where we do that?

File: packages/wdio-allure-reporter/src/index.ts
Comment (christian-bromann) [reply packages/wdio-allure-reporter/src/index.ts:64]: If we define `capabilities` as 
```js
private _capabilities: WebDriver.DesiredCapabilities;
```
we also need to update where we use it to `this._capabilities`

File: packages/wdio-allure-reporter/src/index.ts
Comment (hieuxlu) [inline packages/wdio-allure-reporter/src/index.ts:343]: @christian-bromann This line here uses hook as test

File: packages/wdio-allure-reporter/src/index.ts
Comment (hieuxlu) [reply packages/wdio-allure-reporter/src/index.ts:64]: This is not possible will will result in more ugly code because `super` must be the first call if we have default assignment outside of constructor, and we want `outputDir` to be `allure-results` by default

File: packages/wdio-allure-reporter/src/index.ts
Comment (hieuxlu) [reply packages/wdio-allure-reporter/src/index.ts]: allure, capabilities and isMultiremote is used in unit test and can not be made private. I hope it's ok to leave it as is.

File: packages/wdio-allure-reporter/src/index.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/src/index.ts:46]: Can we move this into a `types.ts` as we have to eventually expose this to users.

File: packages/wdio-allure-reporter/src/index.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/src/index.ts:51]: These should be private as well. In unit test access them via `reporter['_isMultiremote']`

File: packages/wdio-allure-reporter/src/index.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/src/index.ts:162]:
```suggestion
    setCaseParameters(cid: string) {

```

File: packages/wdio-allure-reporter/src/index.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/src/index.ts:371]: can we type these and the following?

File: packages/wdio-allure-reporter/src/utils.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/src/utils.ts:14]:
```suggestion
export const getTestStatus = (test: TestStats | HookStats, config: WebdriverIO.Config): Allure.Status => {

```

File: packages/wdio-allure-reporter/tests/compoundError.test.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/tests/compoundError.test.ts:5]:
```suggestion
    let e1: Error
    let e2: Error

```

File: packages/webdriver/webdriver.d.ts
Comment (christian-bromann) [inline packages/webdriver/webdriver.d.ts:408]: This has been added in a recent PR: https://github.com/webdriverio/webdriverio/blob/master/packages/webdriver/webdriver.d.ts#L567-L589
```suggestion
```

**Review by christian-bromann — COMMENTED** [review]
This is shaping up quite nicely, some last comments

File: packages/wdio-allure-reporter/src/index.ts
Comment (hieuxlu) [reply packages/wdio-allure-reporter/src/index.ts:162]: cid type in `SuittStats` is `string | undefined`. This is a short hand.

File: packages/wdio-allure-reporter/src/index.ts
Comment (hieuxlu) [reply packages/wdio-allure-reporter/src/index.ts:371]: Personally, I don't think it's worth it. These are event listeners, adding an interface would bring no type-checking value because process.emit and process.on will always use anonyous type. Regardless, I think these event handlers should be private

Comment (hieuxlu) [reply packages/wdio-allure-reporter/src/index.ts:371]: The way it is used in unit test also make it really troublesome to define all those typings, with no real benefits
```
    it('should do nothing if no tests run', () => {
        const reporter = new AllureReporter()
        expect(reporter.addLabel({})).toEqual(false)
        expect(reporter.addStory({})).toEqual(false)
        expect(reporter.addFeature({})).toEqual(false)
        expect(reporter.addSeverity({})).toEqual(false)
        expect(reporter.addIssue({})).toEqual(false)
        expect(reporter.addTestId({})).toEqual(false)
        expect(reporter.addEnvironment({})).toEqual(false)
        expect(reporter.addDescription({})).toEqual(false)
        expect(reporter.addAttachment({})).toEqual(false)
        expect(reporter.startStep('test')).toEqual(false)
        expect(reporter.endStep('passed')).toEqual(false)
        expect(reporter.addStep({})).toEqual(false)
        expect(reporter.addArgument({})).toEqual(false)
    })
   ```

**Review by christian-bromann — APPROVED** [review]
👍 thank you so much, fantastic job!

**Pull request merged**
